src/vmm/src/builder.rs

```diff
@@ -763,39 +767,558 @@ pub mod tests {
+        let vcpu_vec = create_vcpus_aarch64(
+            &vm,
+            &vcpu_config,
+            &guest_memory,
+            entry_addr,
+            TimestampUs::default(),
+        )
         .unwrap();
+        assert_eq!(vcpu_vec.len(), vcpu_count as usize);
+    }
+
+    #[test]
+    fn test_attach_net_devices() {
+        let mut event_manager = EventManager::new().expect("Unable to create EventManager");
+        let mut vmm = default_vmm();
+
+        #[cfg(target_arch = "x86_64")]
+        setup_interrupt_controller(&mut vmm.vm).unwrap();
+
+        #[cfg(target_arch = "aarch64")]
+        setup_interrupt_controller(&mut vmm.vm, 1).unwrap();
+
+        // test create network interface
+        let network_interface = NetworkInterfaceConfig {
+            iface_id: String::from("netif"),
+            host_dev_name: String::from("hostname"),
+            guest_mac: None,
+            rx_rate_limiter: None,
+            tx_rate_limiter: None,
+            allow_mmds_requests: false,
+        };
+
+        let mut network_interface_configs = NetworkInterfaceConfigs::new();
+        network_interface_configs.insert(network_interface).unwrap();
+
+        assert!(
+            attach_net_devices(&mut vmm, &network_interface_configs, &mut event_manager).is_ok()
+        );
+
+        // We can not attach it once more.
+        assert!(
+            attach_net_devices(&mut vmm, &network_interface_configs, &mut event_manager).is_err()
+        );
+    }
+
+    #[test]
+    fn test_attach_block_devices() {
+        let mut event_manager = EventManager::new().expect("Unable to create EventManager");
+
+        // Use Case 1: Root Block Device is not specified through PARTUUID.
+        {
+            let mut vmm = default_vmm();
+
+            #[cfg(target_arch = "x86_64")]
+            setup_interrupt_controller(&mut vmm.vm).unwrap();
+
+            #[cfg(target_arch = "aarch64")]
+            setup_interrupt_controller(&mut vmm.vm, 1).unwrap();
+
+            let block_file = TempFile::new("").unwrap();
+            let root_block_device = BlockDeviceConfig {
+                drive_id: String::from("root"),
+                path_on_host: block_file.as_path().to_path_buf(),
+                is_root_device: true,
+                partuuid: None,
+                is_read_only: true,
+                rate_limiter: None,
+            };
+
+            let mut block_dev_configs = BlockDeviceConfigs::new();
+            block_dev_configs.insert(root_block_device).unwrap();
+
+            let res = attach_block_devices(&mut vmm, &block_dev_configs, &mut event_manager);
+            assert!(res.is_ok());
+            assert!(vmm.kernel_cmdline.as_str().contains("root=/dev/vda ro"));
+            assert!(vmm
+                .mmio_device_manager
+                .get_device(DeviceType::Virtio(TYPE_BLOCK), "root")
+                .is_some());
+        }
+
+        // Use Case 2: Root Block Device is specified through PARTUUID.
```

Comment:
Nit: Can you please add a short use case similar to this one, but with `is_read_only` field `true` to cover one more line (line 635) :D? I tried to change the other use cases but it resulted in other lines uncovered :(.

Reply:
Addressed this line. @lauralt , take a look again, please.